### ISSUE TITLE
Credit card error notifications for missing/expired information

### ISSUE THREAD

**Issue by kdjstudios** [kdjstudios/SABillingGitlab#1448]
In GitLab by @kdjstudios on Apr 22, 2019, 13:05

**Submitted by:** "Gabrielle Wachter" <gabrielle.wachter@answernet.com>

**Helpdesk:** http://www.servicedesk.answernet.com/profiles/ticket/2019-04-22-21666/conversation

**Server:** Internal

**Client/Site:** Portland (All)

**Account:** NA

**Issue:** 

We’ve discovered that we have an issue in Portland regarding their auto-pay credit cards. It seems that most, if not all, of the cards had expired and did not go through the process for the last few months. Can you please tell me if there is a notification process in place, any sort of error message, for these types of issues (IE; expired card information, missing information, etc.) when processing credit cards through SAB? If not, is there any way we could add some kind of notification in SAB that alerts the sites when they try to run cards with these type of issues and/or prevents them from moving forward unless corrected?

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Apr 22, 2019, 13:21

I wrote:

> Thank you for reaching out. First I do have one concern as you mention auto pay, however, SA Billing does not have an auto pay feature currently. So I just want to confirm you are processing these in SA Billing and not in possibly in another software. SA Billing does have a Credit Card Batch Payment (report) that allows for a user to manually process multiple accounts payments via one screen, maybe that is what you are referencing. If so, the report does provide receipts and status of each payment when processed and I would expect the user processing them should easily be able to distinguish failed payments. 
> 
> In regards to SA Billing's ability to save CC info; this information is stored in what is refereed to as a 'Vault' on the CC Gateway processing provider and not in SA Billing. If I recall correctly, this information is encrypted as soon as it is stored. Also SA Billing should not allow for saving a CC if there is missing information. We will have to see if the SA Billing support team is able to pull information such as expired cards. If so, we can then have them estimate the development time to create a process to check these and notify operations accordingly. 
> 
> I do see you have also included cash sheet on this email too. I am curious if they possibly have access via the CC Gateway to pull that information or receive notifications from there for expired cards. That might be a quicker and easier way. 

@smasih @amishra @pchaudhary 

First, may we check if we are able to pull saved CC info. If so, may we please have an rough estimate on creating a process to check for expired cards or cards with missing information? I would feel this would be best done via a report or possibly an automated email that gets sent out. We will want to further discuss this I am sure with @tim.traylor and Gary.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Apr 26, 2019, 06:55

@kdjstudios 

Thank you for the update.

We have spent approximate 2.5hrs to finding how to get the expired CC cards but didn't able to find anything yet.

Please suggest to me.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Apr 26, 2019, 11:48

Marking this for discussion. Lets bring this up on our next meeting with @tim.traylor

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Apr 30, 2019, 05:22

marking as an enhancement as well

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on May 2, 2019, 10:14

@amishra @smasih @pchaudhary 

We will want to look further into the ability to retrieve the CC expiration date, or to store it in compliance.

I found the following old topic that discussed the concerns: https://community.developer.authorize.net/t5/The-Authorize-Net-Developer-Blog/Notifying-Users-Their-Credit-Card-Is-About-to-Expire-Without-PCI/ba-p/8025

Posted in the zoom meeting on the compliance way to do this: https://stackoverflow.com/questions/44630367/is-last-4-digits-of-credit-card-and-expiry-date-storage-allowed-in-pci-dss

Overall, from my understanding, it does appear we should be able to store this information in compliance. I would think if we could have the following estimates, we can then have @tim.traylor and Gary approve which is the better path or if all are needed. 

1. Develop the ability to track expiration dates in SAB in compliance.
2. Send out a notification for each site that includes all cards expiring in the next month, on the first of each month.
3. Develop a report snapshot that contains all cards expiring in the upcoming month, on the first of each month. 

I figure if we do this on a monthly recurrence then ops will only need to check once a month, and we dont have to check every day, etc.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on May 13, 2019, 04:49

@kdjstudios 

Thank you for the update.

|Size|Lower Time Bound|Variability|Upper Time Bound|Testing|Time for Testing| Time for Dev & Testing|
|-|-|-|-|-|-|-|
|Large-1|8|Worked with Concepts, implementation ok|12|Low Impact|3.0|15|

**Bifurcation of the estimate**

* Create a new field of the Account modal for saving the expiry date.
* Create a script for checking to expire in the next month.
* Create a cronjob to send an email.
* Show on the snapshot report.
* Testing

Note: Please post the verbiage of body & subject of the email and also let us know who will be the recipients and the sender.
 

Thanks.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on May 13, 2019, 07:35

@pchaudhary 

How are we planning to get all expiry dates ?? Is there a Authorize API or service

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on May 13, 2019, 08:01

@smasih Sumeet, I believe the getCustomerPaymentProfile method in the CIM API can be used to return the expiration date.  You have to specifically pass an indicator to request the unmasked expiration date.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 5, 2019, 10:05

@amishra @smasih @tim.traylor 

Marking back for estimate. @tim.traylor has suggested a possible better way.

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Jun 7, 2019, 09:52

Breakdown:
* We will have a new collection of CC expiration info. This will include the account, the last 4 of the CC number, the expiration date.
* Any time a new CC is entered via the external customer payment portal or the internal screens (and they request to save the card info), we will create or overwrite an entry in the new collection.

Initial population:
Create a script to run the getCustomerPaymentProfile method in the Authorize.Net CIM API to return the expiration date and populate the new collection.  This is only done one time.

Reporting:
* We will create a 'standard' type of report that can be run from the UI.
* When running the report you can specify how far in the future to show cards expiring (e.g., next 30 days, 60 days, 90 days, etc.). It should pass this 'number of days in the future' into the script for the scope.  It will use today's date as the starting point.
* The report should include the account info, the account/customer contact info, the last 4 of the CC and the expiration date.
* The report will also have an 'Export to PDF' option.

Scheduling:
* We will create cron task that will run the first of the month.
* The job will execute the 'export to PDF' function for each site, specifying 45 days as the 'number of days in the future'.
* As each file is created it will email it to the site email address.

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Jun 11, 2019, 06:25

@kdjstudios Pls review the process w/Gary before we get the revised estimate.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 11, 2019, 09:30

I was able to review this with Gary today and he inquired if this information was directly available via a report in Authorize.Net. I believe it was Richie who stated that we had researched this and that it was not, due to it being encrypted. 

Gary approves that we will need to have a report to the users to run and also for a notification on the begging of the month of upcoming expired cards. Marking this for estimate.

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Jun 11, 2019, 12:53

![image](/uploads/f626af39313daeb1ec999875510e1533/image.png)

So the price was reasonable, but they don't cover AMEX.  So we do need to do it for ourselves.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Aug 20, 2019, 09:26

This was brought back up today during discussion on #1551 and Looks like we have been waiting on an estimate for two months. Gary mentioned when we thought we would have this by.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Aug 23, 2019, 08:07

Secondary ticket to #1274 please keep all updates posted on the parent ticket.

**Comment by kdjstudios** [comment]
In GitLab by @tim.traylor on Aug 30, 2019, 06:32

@kdjstudios Hi Kyle - I don't see anything for me to approve here...

@smasih Sumeet, I believe we'res till waiting on the estimate of the functionality I posted on this comment: http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1451#note_41821

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Aug 30, 2019, 09:51

@tim.traylor 

noted thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Aug 30, 2019, 14:26

@tim.traylor 

This became a child ticket to #1274, which is where all the updates and estimate will be placed at. We simply on child tickets keep the same labels so they do not get lost. 

I will also notate on #1274 to confirm that the request in http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1451#note_41821 is part of the estimate.

**Comment by kdjstudios** [comment]
In GitLab by @jredfield on Sep 17, 2019, 06:08

This is a credit card expiry notification feature and yet to be estimated (specs already provided)

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Sep 20, 2019, 01:36

@kdjstudios @jredfield 

> This became a child ticket to #1274, which is where all the updates and estimate will be placed at. We simply on child tickets keep the same labels so they do not get lost.
>
> I will also notate on #1274 to confirm that the request in http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1451#note_41821 is part of the estimate.

I don't think if these two tickets a related as, #1274  is a CC validation ticket, and this ticket is for running a scheduled job and a report. 

## Developer Estimate
|Size|Lower Time Bound|Variability|Upper Time Bound|Testing|Time for Testing| Time for Dev & Testing|
|-|-|-|-|-|-|-|
|Medium-1|4|Worked with Concepts, implementation ok|6|Low Impact|1.5|7.5|

### Question

* This functionality is for only accounts with saved cards (ccof) right?

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Sep 20, 2019, 08:40

@smasih @amishra @jredfield @tim.traylor 

Why was this not part of #1274 as requested in http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1451#note_46612. All of this development was supposed to be done on the same ticket and at the same time with the same estimate per http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1274#note_47401. I even posted a confirmation in http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1274#note_47617; if that was incorrect why did no one mention it? Come on guys we need better communication!!!


It doesn't matter if it was 'different development' they all deal with the CC payments, thus why they were all supposed to be done at the same time. We do not want to spread these updates out over multiple releases. They needed to all be tested in conjunction to ensure they properly worked together.

Please move this estimate to the parent ticket as originally requested and perform the work there.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Sep 22, 2019, 23:50

@kdjstudios 

I understand the purpose of keeping all CC related changes in one place but, it becomes hard to track time and keep discussions segregated. This is a Child ticket of #1274 not a Duplicate of #1274 so we can post discussion and estimates on this ticket.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Sep 23, 2019, 11:47

@jredfield 

The 7 hour estimate provided does appear to me to be acceptable, as they are creating a report and schedule for it to auto run too. Please mark your approval if it looks good to you.

**Comment by kdjstudios** [comment]
In GitLab by @jredfield on Sep 27, 2019, 05:39

Approved

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Oct 3, 2019, 05:06

assigned to @amishra

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Oct 3, 2019, 06:27

@kdjstudios @jredfield 

Thank you for your approval.

> Reporting:

 > - We will create a 'standard' type of report that can be run from the UI.

 > - When running the report you can specify how far in the future to show cards expiring (e.g., next 30 days, 60 days, 90 days, etc.). It should pass this 'number of days in the future' into the script for the scope.  It will use today's date as the starting point.

 > - The report should include the account info, the account/customer contact info, the last 4 of the CC and the expiration date.

 > - The report will also have an 'Export to PDF' option.


It would be great if you will provide the mockup for the report section.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @jredfield on Oct 3, 2019, 07:15

@amishra You changed this to being in development. Do you need the mockup first?

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Oct 3, 2019, 07:43

@amishra @kdjstudios @jredfield 

although specs are explained here (http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1451#note_41821) but having a mockup would be great, especially when everyone is not clear on how to deal with a reporting ticket.

**Comment by kdjstudios** [comment]
In GitLab by @jredfield on Oct 3, 2019, 07:47

Let's discuss in today's deep dive meeting.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 3, 2019, 09:17

Here is a basic mock up. Let us know if you have any questions or concerns. 


![CC_Expiration_date_report](/uploads/32637d313684d35ede2f2759a68dc429/CC_Expiration_date_report.png)

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Oct 4, 2019, 24:18

@kdjstudios 

Thank you for the mockup.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 7, 2019, 12:54

@amishra 

I forgot to add this in the mock up, But please make the Account name a link to the account page.

**Comment by kdjstudios** [comment]
In GitLab by @rranjan on Nov 14, 2019, 03:33

@kdjstudios @jredfield 

Thank you for the update

* We have setup a branch on merge server for testing the Reporting section.

* We have developed the cron task to run on the first of every month. 
  Please specify the time at which we need to run it

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 14, 2019, 10:15

@rranjan @jredfield 

The report looks great but we have a few issues:

1. There seems to be an issue with the customer name column. As you can see the CSV populated, but the web UI and PDF are blank.
2. The CSV appears to have the last 4 and email columns switched. 

![image](/uploads/a77ac192a9337c1b9f12801246e1e06f/image.png)

[credit_card_expirations_11_14_2019.pdf](/uploads/db891391a1074e9f28ead61f1eba5f0e/credit_card_expirations_11_14_2019.pdf)

[credit_card_expirations_11_14_2019.csv](/uploads/eeee9d35e1760addc3019e564232ec2d/credit_card_expirations_11_14_2019.csv)

In regards to the email time, may we please have this ran at 6am ET on the first of every month. Also I presume we will need to setup the email template in the admin section too? May we please have the template wording in the email be as listed below. Lastly, I am uncertain if this was part of the request, but I do not believe we should be sending the below email if there are no CC expiring. 

```
From: Support@sabilling.com
To: [Site Email]
BCC: apperrors@sahosted.com
Subject: SA Billing - Upcoming Saved Credit Cards Expiring Soon
Body:
Hi,

Please see the attached file, which contains a list of the saved credit cards in SA Billing that will be expiring within 45 days. We recommend to reach out to those customers and retrieve updated credit card information to ensure there is no missed payments. 

Thanks, SA Billing Support.
```

**Comment by kdjstudios** [comment]
In GitLab by @rranjan on Nov 15, 2019, 06:28

@kdjstudios @jredfield 

Thank you for the update

We have fixed the issue and now showing correctly and setup branch on the merge server for your testing.

Below is the format for the email attachment [Portland__OR.pdf](/uploads/132e752f3e793c0deb739d28adac2adf/Portland__OR.pdf) . **Please check it**

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 15, 2019, 11:21

@rranjan 

Thank you for the quick updates. All appears to be working on the report and the exports now.

In regards to the email, I am not seeing this setup in the email templates section in the admin yet?

![image](/uploads/07ee53fd7f263e295acc617c6183c784/image.png)

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Nov 15, 2019, 11:57

@kdjstudios @jredfield 

Thank you for the update.

> In regards to the email, I am not seeing this setup in the email templates section in the admin yet?

Yes, not yet we will do it.  

@rranjan wants to confirm that the pdf format looks like [pdf](http://gitlab.aavaz.biz/AnswerNet/SABilling/uploads/132e752f3e793c0deb739d28adac2adf/Portland__OR.pdf) while sending an email of account details for each site..

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 15, 2019, 12:10

@amishra @rranjan 

Thank you for the clarification. No the PDF attachement does not, it should be exactly the same PDF that gets generated by the report. 

[credit_card_expirations_11_15_2019__1_.pdf](/uploads/9e5e02f7830a40f5f97a22e61b13fd79/credit_card_expirations_11_15_2019__1_.pdf)

**Comment by kdjstudios** [comment]
In GitLab by @ranand on Nov 18, 2019, 05:43

@kdjstudios @jredfield 

Thank you for the update.

We have updated the credit card expiration Pdf format. Please see the attached Pdf. [credit_card_expirations.pdf](/uploads/690951373a273fd41796b5891ba01a9d/Sacramento__Ca._.pdf)

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 18, 2019, 11:46

@ranand 

Thank you, that looks great. Please let us know when the email template is done too.

**Comment by kdjstudios** [comment]
In GitLab by @ranand on Nov 18, 2019, 22:18

@kdjstudios @jredfield 

The email template is also done, please see the attached image. 

![image](/uploads/4d31fb02e3ac2075808b7dbd3f2e7d5b/image.png)

This is how the email template is looking.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 19, 2019, 12:21

@ranand 

Thanks! Looks great marking this for announcement.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Nov 23, 2019, 05:01

@kdjstudios @jredfield 

Thank you for the update.

We have successfully released on both of the production servers.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 23, 2019, 11:39

report looks good.we will need to wait until the 1st of dec. to confirm that the email is sending.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Dec 2, 2019, 08:53

We got the emails, all is good. marking to close.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Dec 2, 2019, 08:53

closed
